Purposed Version 1.1.0 - Temperature Sensor Support

iLikePieTM opened the pull request:
I would like to propose a version 1.1.0 update to the library that adds support for the LSM6DS3's integrated temperature sensor with three new functions:

- `readTemp` outputs the temperature reading in deg. C and returns 1 if successful

- `tempSampleRate` returns the sampling rate of the sensor

- `tempAvailable` returns 1 if there is available data from the temperature sensor


_P.S. first time doing this so don't be too harsh :)_

per1234 commented on examples/SimpleTempSensor/SimpleTempSensor.ino:
```diff
@@ -0,0 +1,45 @@
+  Serial.print(IMU.tempSampleRate());
+  Serial.println(" Hz");
+  Serial.println();
+  Serial.println("Temperature sensor in degrees C");
```
```suggestion
  Serial.println("Temperature reading in degrees C");
```

per1234 commented on examples/SimpleTempSensor/SimpleTempSensor.ino:
```diff
@@ -0,0 +1,45 @@
+}
+
+void loop() {
+  float t;
```
I think it would be nice to have a comment explaining the purpose of this variable. Beginners might expect that `IMU.readTemp()` will return the temperature reading, and not be familiar with the concept of passing arguments by reference.

iLikePieTM replied:
Good idea, I'm just not sure how I would word it. If you have a suggestion I could implement it in the other two examples too.

per1234 replied:
I decided the better place for the explanatory comment would be on the `IMU.readTemperature()` call. I submitted a suggestion for this on that line.

>I could implement it in the other two examples too

I think that would be great, but please do it in a separate pull request to keep this pull request atomic.

per1234 commented on library.properties:
```diff
@@ -1,8 +1,8 @@
 name=Arduino_LSM6DS3
-version=1.0.0
+version=1.1.0
```
```suggestion
version=1.0.0
```
I think it's best to leave the versioning to the repository maintainers. The version is normally bumped right before the release, so it might cause confusion to have it pre-bumped.

per1234 commented on src/LSM6DS3.cpp:
```diff
@@ -31,6 +31,8 @@
 #define LSM6DS3_CTRL7_G            0X16
 #define LSM6DS3_CTRL8_XL           0X17
 
+#define LSM6DS3_OUT_TEMP_L 			   0X20
```
```suggestion
#define LSM6DS3_OUT_TEMP_L         0X20
```
Use spaces for alignment. People can have true tabs set to any width so what looks aligned with your tabs setting is not aligned for someone else's tab setting.

iLikePieTM commented:
> Consider using `readTemperature`, `temperatureAvailable`, `temperatureSampleRate` to make the code more self-documenting. "temp" might be short for "temporary".
> 
> Please add the new function names to keywords.txt. Use a true tab to separate the function name from the identifier:
> https://github.com/arduino/Arduino/wiki/Arduino-IDE-1.5:-Library-specification#keywordstxt-format

Thanks for the feedback!

- I implemented the self-documenting suggestion; I tend to use abbreviations but there's no reason not to be descriptive. I also updated the keywords file.

- Sorry for getting a bit ahead of myself with the revision number changes too :P

per1234 commented on keywords.txt:
```diff
@@ -9,19 +9,22 @@ LSM6DS3	KEYWORD1
+temperatureAvailable  KEYWORD2
 accelerationSampleRate	KEYWORD2
 gyroscopeSampleRate	KEYWORD2
+temperatureSampleRate KEYWORD2
```
```suggestion
readTemperature	KEYWORD2
gyroscopeAvailable	KEYWORD2
accelerationAvailable	KEYWORD2
temperatureAvailable	KEYWORD2
accelerationSampleRate	KEYWORD2
gyroscopeSampleRate	KEYWORD2
temperatureSampleRate	KEYWORD2
```
The Arduino IDE only recognizes the keywords when you use a true tab separator here, not spaces.

iLikePieTM replied:
Sorry about that, I tried to use tab.

per1234 commented on src/LSM6DS3.cpp:
```diff
@@ -176,14 +178,43 @@ float LSM6DS3Class::gyroscopeSampleRate()
+    return 0;
+  }
+
+  t = data[0] / 16 + 25;
```
```suggestion
  t = data[0] / 16.0 + 25;
```
The previous integer math caused the temperature data to be truncated.

iLikePieTM replied:
Good catch!

per1234 commented on examples/SimpleTempSensor/SimpleTempSensor.ino:
```diff
@@ -0,0 +1,45 @@
+  float t;
+
+  if (IMU.temperatureAvailable()) {
+    IMU.readTemperature(t);
```
```suggestion
    // after IMU.readTemperature() returns, t will contain the temperature reading
    IMU.readTemperature(t);
```

cansik commented:
Will this be merged soon?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino-libraries/Arduino_LSM6DS3?pullRequest=12) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino-libraries/Arduino_LSM6DS3?pullRequest=12) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino-libraries/Arduino_LSM6DS3?pullRequest=12) it.</sub>

J1mbo commented:
> @iLikePieTM it seems your PR was approved long time ago, will it be merged soon? 😄 (this is a great feature)

+1!

isyca commented:
very useful to have and hope this gets merged. worked well with Nano 33 IoT

a little suggestion if I may,  to follow LSM6DSOXClass::readTemperature and LSM6DSOXClass::readTemperatureFloat from Arduino_LSM6DSOX that way we have int and float

oliverswitzer commented:
Would love to bump this, since all checks have passed and its been approved. This would be immensely helpful to have on my current project.

iLikePieTM commented:
It's been years and I totally forgot about this lol. I fixed the conflicts with the current master branch, it's ready to merge by someone with write access.

oliverswitzer commented:
@iLikePieTM thanks for the quick response! Really excited to make use of this.

@per1234 any chance this could get a second review / a merge?

per1234 commented:
Hello all. Thanks @iLikePieTM for your pull request and also to @oliverswitzer for your interest in this open source project.

My role in this project is essentially janitorial. I am not in a role to merge changes to the library codebase. One of the projects other maintainers will make that decision.

Regards,
Per

oliverswitzer commented:
Hi @iLikePieTM,

I am trying to use this functionality now, but am consistently getting temperature readings that seem like they're too high considering that the board is sitting at ambient room temperature (75 degrees F):

```
float temp = 0;

...

  if (IMU.temperatureAvailable()) {
    IMU.readTemperature(_lastTemp);
    IMU.readTemperature(temp);
    Serial.println("*********** TEMP: " + String(temp));
  }
```

Output (over a few seconds):
```
********** TEMP: 133.88
*********** TEMP: 130.81
*********** TEMP: 133.38
*********** TEMP: 132.06
*********** TEMP: 131.69
*********** TEMP: 129.25
*********** TEMP: 133.56
*********** TEMP: 132.88
*********** TEMP: 130.56
*********** TEMP: 132.75
*********** TEMP: 130.12
*********** TEMP: 130.69
*********** TEMP: 132.44
*********** TEMP: 133.94
*********** TEMP: 130.69
*********** TEMP: 134.56
*********** TEMP: 130.75
*********** TEMP: 131.94
*********** TEMP: 133.19
*********** TEMP: 130.38
*********** TEMP: 133.38
*********** TEMP: 131.88
*********** TEMP: 132.00
*********** TEMP: 133.44
*********** TEMP: 132.56
*********** TEMP: 131.50
*********** TEMP: 130.81
*********** TEMP: 132.19
```

Any idea why this temp seems so high?
